### DISCUSSION TITLE
Discuss about MMDeploy pre-build.

### DISCUSSION THREAD

**Discussion by grimoire** [open-mmlab/mmdeploy#354] [General]
We plan to create some pre-build package to ease the installation of MMDeploy. Here are some details:

## Environment and Platform

**environment**

- linux-x86_64 （cpu/cuda11.x)
- windows-x86_64 (cpu/cuda11.x)
- Jetson

**backend**

- TensorRT 8
- ONNX Runtime 1.8

We will support these combinations. More systems and backend would be added in the future ( if needed).

## Plan

**Model Converter** and **SDK** would be built in two different packages.

### Model Converter

Most of the code in the model converter is python, so we plan to align the pre-build package with PyTorch.

- Decouple `export IR` and `convert backend model` is a good choice. Since user might want to deploy their model in some edge device (jetson for example).It is not easy to install some third-party library on these devices. Instead, we can export ONNX on PC and convert TensorRT model on jetson device with **NO** third-party dependency.
- All compiled libs (custom ops, export tools etc) should be placed in `mmdeploy/lib` so they can be packed in the same wheel with the  python code.
- Some tools can be mapped to `mmdeploy` so we can use them like `python -m mmdeploy deploy ...`.
- The wheel will be placed in our server (or somewhere else that can be access public), user can install the package with `pip install mmdeploy=={version} -f https://<url_to_wheels>`

### SDK

The behavior of SDK looks like an inference engine, so the package would be aligned with TensorRT pre-build package.

- SDK would be packed as `tar` or `zip` package, which includes heads, libraries, python API, and anything that might be necessary. It would be published after each release of MMDeploy.
- SDK could be used independently of the model converter.
- The python API (I mean `mmdeploy_python`) will be another wheel, which is placed in `<package_dir>/python`, just like TensorRT. User can optionally install the API.

I have made a private repo [mmdeploy_builder.py](https://github.com/grimoire/mmdeploy_builder) to do the model converter build. It is not well tested, and is buggy(I guess). SDK build tools have not yet been developed.

There are two PR related:

- [ ] https://github.com/open-mmlab/mmdeploy/pull/273 which tries to remove the dependency of PyTorch and any other third-party libs.
- [ ] https://github.com/open-mmlab/mmdeploy/pull/347 move the library to the `mmdeploy/lib` so we can pack them together with python code.

I plan to close #273 since it is an old PR with a lot of conflict with the current master. It would be done in #347 again.

There is still a lot to do:

- [ ] SDK build tools, include heads, libs, cmake, python API (most of them can be provided by the installation besides the python API). @PeterH0323 
- [ ] Decouple `export IR` and `convert backend model`. (remove third-party dependency, make sure we can convert TensorRT model on the jetson platform without installing PyTorch)
- [ ] windows support.
- [ ] bug fix
- [ ] Anything I might forget to mention.
- [ ] build script or tools to ease the installation.

@tehkillerbee @PeterH0323 Please feel free to choose the tasks (or is there any task I missed?). Let's work together to make MMDeploy Better!

**Comment by tehkillerbee** [comment]
@grimoire So just to understand, the plan is to have a tool that can build MMDeploy + SDK for various platforms and with various backends? To allow pre-build releases of MMDeploy that can be easily downloaded by the user?

Initially, I thought a single build script would be used, similar to how it is done with ppl.cv. This might be the easiest solution but it requires the user to build locally. This is what I ended up doing myself.

If we want to do pre-build packages, my main concern is how to handle building for other platforms - eg. aarch64, as I assume we would crosscompile the packages on a non aarch64 build server?

I have previously tried crosscompiling ppl-cv, MMDeploy and the dependencies, but I did not manage to get it working properly due to the way the dependencies are detected. Instead, I wrote a bash script that handles building and installing all dependencies (including spdlog, pplcv) and MMDeploy SDK on the Jetson itself. Then I was able to crosscompile my custom inference tools on my dev. machine using the Jetson rootfs containing the pre-built MMDeploy SDK libs.

Decoupling `IR` and `convert backend model` to avoid pytorch on Jetson sounds like a good idea, that would simplify install steps if you only need to do TensorRT model conversion on Jetson. Then all other steps can be done in advance on the development machine.

By the way, how do we handle misc. dependencies that are not installed by MMDeploy pre-build package? Eg. mmcv-full and other python packages that are needed for mmdeploy? Can they be installed automatically as part of the installation of MMDeploy wheel?

**Comment by PeterH0323** [comment]
> By the way, how do we handle misc. dependencies that are not installed by MMDeploy pre-build package? Eg. mmcv-full and other python packages that are needed for mmdeploy? Can they be installed automatically as part of the installation of MMDeploy wheel?

Yes. We will install the dependencies which we need when we install the mmdeploy wheel package.

**Comment by PeterH0323** [comment]
> If we want to do pre-build packages, my main concern is how to handle building for other platforms - eg. aarch64, as I assume we would crosscompile the packages on a non aarch64 build server?

We need to use the aarch64 to build it and then relesae it, just like some pre-build package need to choose `x86` or `aarch64`.

**Comment by grimoire** [comment]
@tehkillerbee 
A build script is a good idea  (Actually I am planning to add an interact script, or even install tools with a panel, which we can fill whatever dependency we need then click `next`). I always want to add this. It can also be part of this plan (now).
But still, people might want some packages that they can download and use without any work to do. We do not have to cover all platforms. They can build it with source code by themself if their platforms has not been pre-built yet. (Just like mmcv, we can install the pre-built one or build from source). 
We do not need to worry about cross-compile for now. The pre-build for jetson can be built on the device for now until we figure out how to build it on x86-64.
As for the `misc` you mentioned. We can add part of it into the requirements, but I guess some dependency should be installed manually(such as the installation requires login to their main page). That is not a big deal since user should always install these package whatever they use pre-built package or not.

**Comment by PeterH0323** [comment]
> So just to understand, the plan is to have a tool that can build MMDeploy + SDK for various platforms and with various backends? To allow pre-build releases of MMDeploy that can be easily downloaded by the user?

Yes, cuz some users have trouble when they deploy into some device. At the same time, they need to install all mmdeploy dependencies step by step according to the doc which is little harder for some users. I think this pre-build package will help more users to use mmdeploy more fluently.

**Comment by PeterH0323** [comment]
> There is still a lot to do:
> 
> * [ ]  SDK build tools, include heads, libs, cmake, python API (most of them can be provided by the installation besides the python API).
> * [ ]  Decouple `export IR` and `convert backend model`. (remove third-party dependency, make sure we can convert TensorRT model on the jetson platform without installing PyTorch)
> * [ ]  windows support.
> * [ ]  bug fix
> * [ ]  Anything I might forget to mention.
> 
> Please feel free to choose the tasks (or is there any task I missed?). Let's work together to make MMDeploy Better!

Hi @grimoire , I want to pick `SDK build tools, include heads, libs, cmake, python API (most of them can be provided by the installation besides the python API).` 😄

**Comment by tehkillerbee** [comment]
@grimoire I could probably assist with the build script, since I have already worked on something like that.
I can also show the simple bash script I wrote for installing dependencies + MMDeploy SDK.

`export IR and convert backend model` sounds like quite a large task and would probably require some pointers if I were to look at that. Do you have any suggestions on where to start with that task?

**Reply by grimoire** [reply]
Cool, let's start with your build script.

`export IR and convert backend model` include `code decoupling` and `new APIs designing`. We just need to make sure tools like `mmdeploy.backend.tensorrt.create_trt_engine` won't depend on PyTorch or MMCV. `new APIs` is designing and would be released in the future.

**Reply by tehkillerbee** [reply]
@grimoire 

So far, this is how it looks. Please note it is a quick & dirty install script at this point. But it works on both x86_64 and aarch. 

I also have another script for installing MMDeploy and misc. dependencies. That script takes care of setting up the python venv as well.

I experimented with packing the build contents to a tar.gz file, but it does not pack everything, unless I install to a temp output folder first. 

```sh
#!/bin/bash
# build_mmdeploy.sh
#   Date: 08-03-2022
#   Author: Johannes Linde, jo@dralle.dk
#   Run this script to build MMDeploy SDK and dependencies (ppl.cv)
#   Note: This script does NOT support cross compilation yet. Run on Jetson platform and copy libs for later crosscompilation
#   Note: Use build_mmdet.sh to setup Python Venv before running this script
#
BUILD_TYPE="Release"
BUILD_ARCH=$(uname -i)
WITH_PYTHON=1
WITH_CLEAN=1
# Get path to script to ensure script runs from 'logends' root
WORKING_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"

PPLCV_DIR=${WORKING_DIR}/ppl.cv
MMDEPLOY_DIR=${WORKING_DIR}/MMDeploy
PPLCV_VER="0.6.2"
MMDEPLOY_VER="0.3.0"

appargument1=$1
#appargument2=$2

echo_green() {
  if [ -n "$1" ]; then
    echo "$(tput setaf 10)$1$(tput sgr 0)"
  fi
}
echo_red() {
  if [ -n "$1" ]; then
    echo "$(tput setaf 1)$1$(tput sgr 0)"
  fi
}
echo_blue() {
  if [ -n "$1" ]; then
    echo "$(tput setaf 4)$1$(tput sgr 0)"
  fi
}
contains_element () {
  local e match="$1"
  shift
  for e; do [[ "$e" == "$match" ]] && return 0; done
  return 1
}

pplcv() {
  ## ppl.cv - install in /usr/local
  cd ${WORKING_DIR}
  echo_blue "checking out '${PPLCV_DIR}' pkg..."
  if [ -d "${PPLCV_DIR}" ]; then
      echo_green "Already exists! Checking out the requested version..."
  else
      git clone https://github.com/openppl-public/ppl.cv.git ${PPLCV_DIR}
  fi
  cd ${PPLCV_DIR}
  git pull
  git checkout tags/v${PPLCV_VER}
  mkdir build -p && cd build
  cmake -DHPCC_USE_CUDA=ON -DCMAKE_INSTALL_PREFIX=/usr/local .. && make -j${processor_num} && sudo make install
  sudo ldconfig
  # pack as tar.gz file
  cd ..
  tar -zcvf ${WORKING_DIR}/pplcv_${PPLCV_VER}_cuda-${BUILD_ARCH}-build.tar.gz build/
}

mmdeploy(){
  ## h5py (Required by mmdeploy)
  ## h5py not directly supported by jetson and must be built/installed manually
  sudo apt-get install pkg-config libhdf5-10* libhdf5-dev -y
  sudo pip3 install Cython
  sudo env H5PY_SETUP_REQUIRES=0 pip3 install -U h5py==2.9.0

  ## Check if ${TENSORRT_DIR} env variable has been set
  if [ -d "${TENSORRT_DIR}" ]; then
    echo_green "TENSORRT_DIR env. variable has been set ${TENSORRT_DIR}"
  else
    echo_red "TENSORRT_DIR env. variable has NOT been set."
    if [[ "$BUILD_ARCH" == aarch64 ]]; then
      echo_green "Added TENSORRT_DIR to env. Please restart script!"
      echo 'export TENSORRT_DIR="/usr/include/'${BUILD_ARCH}'-linux-gnu/"' >> ${HOME}/.bashrc
      source ${HOME}/.bashrc
      exec bash
    else
      echo_red "Please Install TensorRT, add TENSORRT_DIR to .bashrc before running this script"
      exit
    fi
  fi

  # Determine TensorRT version and set paths accordingly
  echo "Checking TensorRT version...Please verify the detected versions below:"
  cat ${TENSORRT_DIR}/NvInferVersion.h | grep NV_TENSORRT 
  cat /usr/include/${BUILD_ARCH}-linux-gnu/NvInferVersion.h | grep NV_TENSORRT
  read -p "Is TensorRT >=8.0.1.6 installed? (Always installed on Jetson) (y/n)" -n 1 -r
  echo
  if [[ $REPLY =~ ^[Nn]$ ]]
  then
      echo "Error: You must install TensorRT before installing MMDetection!"
      exit
  fi

  ## mmdeploy SDK - install in /usr/local
  cd ${WORKING_DIR}
  echo_blue "checking out '${MMDEPLOY_DIR}' pkg..."
  if [ -d "${MMDEPLOY_DIR}" ]; then
    echo_green "Already exists! Checking out the requested version..."
  else
    git clone https://github.com/open-mmlab/mmdeploy.git ${MMDEPLOY_DIR}
  fi
  cd ${MMDEPLOY_DIR}
  git pull
  git checkout tags/v${MMDEPLOY_VER}
  if [[ $WITH_PYTHON -eq 1 ]]
  then
    pip install -e .
  fi
  rm -r build
  mkdir build -p && cd build
  cmake .. \
    -DMMDEPLOY_BUILD_SDK=ON \
    -DCMAKE_INSTALL_PREFIX=/usr/local \
    -DCMAKE_CXX_COMPILER=g++ \
    -Dpplcv_DIR=/usr/local/lib/cmake/ppl \
    -DMMDEPLOY_TARGET_DEVICES="cuda;cpu" \
    -DMMDEPLOY_TARGET_BACKENDS=trt \
    -DMMDEPLOY_CODEBASES=all
    #-DTENSORRT_DIR=${TENSORRT_DIR} We will rely on env. variable set by system
  cmake --build . -- -j$(nproc) && sudo cmake --install .
  sudo ldconfig
  # pack as tar.gz file
  cd ..
  tar -zcvf ${WORKING_DIR}/mmdeploysdk_${MMDEPLOY_VER}_${BUILD_ARCH}-build.tar.gz build/
  # Unpack as tar -zxf mmdeploysdk_*.tar.gz --directory MMDeploy-aarch64

  # add MMDeploy_DIR to bashrc
  #echo -e 'export MMDeploy_DIR='${WORKING_DIR}'/MMDeploy/build/install/lib/cmake/MMDeploy' >> ${HOME}/.bashrc
  #source ${HOME}/.bashrc

  ## mmdeploy examples
  #export LD_LIBRARY_PATH=${WORKING_DIR}/MMDeploy/build/install/lib
  cp -r ${WORKING_DIR}/MMDeploy/demo/csrc ${WORKING_DIR}/MMDeploy/build/example
  cd ${WORKING_DIR}/MMDeploy/build/example
  rm -r build
  mkdir build -p && cd build
  #cmake -DMMDeploy_DIR=${WORKING_DIR}/MMDeploy/build/install/lib/cmake/MMDeploy ..
  cmake ..
  make all
}

all() {
  # build all
  pplcv
  mmdeploy
}

#####
# build selected
package_list=(
  "all"
  "pplcv"
  "mmdeploy"
)

# check input argument
if contains_element "$appargument1" "${package_list[@]}"; then
  echo_green "Build and install '$appargument1'..."
else
  echo_red "Unsupported argument '$appargument1'. Use one of the following:"
  for i in ${package_list[@]}
  do
    echo $i
  done
  exit
fi

# setup 
cd ${WORKING_DIR}

if [[ $WITH_PYTHON -eq 1 ]]
then
  source venv-mmdet/bin/activate
fi

# remove all build files
if [[ $WITH_CLEAN -eq 1 ]]
then
  sudo rm -r ${PPLCV_DIR}/build
  sudo rm -r ${MMDEPLOY_DIR}/build
fi
# mmdeploy + dependencies (OpenCV, spdlog is assumed to be installed)
git submodule update --init --recursive

$appargument1
```

**Reply by grimoire** [reply]
Great! We just need to add more detail and do more tests.
I think it would be better if we have more interactable options rather than environment variables. User can enter their own environment or just press enter to use the default config.

**Reply by grimoire** [reply]
Hi, would you like to create a PR about the build script? We do not have a detailed design of the build script. Your PR would be a good start.

**Reply by tehkillerbee** [reply]
@grimoire Sure, should I change anything about the script? Where do you suggest the script should be added? Root folder?

I will remove the author information and add Copyright (c) OpenMMLab. All rights reserved.

**Reply by grimoire** [reply]
Root folder is OK. And I think you can keep the author information if you want. 
You can try adding other backend support in the script, or just push the script and discuss it with other reviewers.

**Reply by tehkillerbee** [reply]
@grimoire @PeterH0323 Since the latest python build script also builds the SDK, does it make sense to add this bash script as well?

**Reply by grimoire** [reply]
Sounds like a good idea.

**Reply by tehkillerbee** [reply]
@grimoire Ok, I will add the script if you think it is a good idea. Should I add it to your repo or MMDeploy?

EDIT: Maybe it would make sense to use the .yaml config files for the bash script as well? Only problem is the way pplcv paths are handled by my script as I build ppclv together with mmdeploy.

**Reply by tehkillerbee** [reply]
I have created PR #399 for further discussion with other reviewers, I hope this is OK with you.
